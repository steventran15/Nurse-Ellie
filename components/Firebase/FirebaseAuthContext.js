import React, { createContext } from 'react';

export const FirebaseAuthContext = createContext();