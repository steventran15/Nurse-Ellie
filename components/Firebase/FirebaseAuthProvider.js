import React, { useState, useEffect } from 'react';
import { firebase } from './config';
import { FirebaseAuthContext } from './FirebaseAuthContext';

const FirebaseAuthProvider = ({ children }) => {

    const [currentUser, setUser] = useState(null);

    useEffect(() => {
        let unsubscribe = firebase.auth().onAuthStateChanged(user => {
            if (user) {
                setUser(user);
            }
        })

        return () => {
            unsubscribe();
        }

    }, []);

    return (

        <FirebaseAuthContext.Provider value={{ currentUser }}>
            {children}
        </FirebaseAuthContext.Provider>

    );
};

export {
    FirebaseAuthProvider
}