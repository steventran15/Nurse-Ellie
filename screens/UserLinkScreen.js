import React, { useEffect, useState, useContext } from 'react';
import { View, Text, TouchableOpacity, StyleSheet, Dimensions, TextInput, KeyboardAvoidingView, Keyboard, Alert } from 'react-native';
import * as Animatable from 'react-native-animatable'

import Background from '../components/background';

import NurseEllieConnectLogo from '../assets/images/ellie-connect-logo.svg';
import NurseEllieLogo from '../assets/android/drawable-hdpi/entry-logo.png';
import HP_Btn from '../assets/images/nurse-unselected-icon.svg';
import HP_BtnSelected from '../assets/images/nurse-selected-icon.svg';
import FamilyFriendBtn from '../assets/images/familyfriend-unselected-icon.svg';
import FamilyFriendBtnSelected from '../assets/images/familyfriend-selected-icon.svg';
import CloseBtn from '../assets/images/close-button.svg';

import PatientStyles from '../styles/PatientStyleSheet';

import Modal from 'react-native-modal';

import QRCode from 'react-native-qrcode-svg';
import { firebase } from "../components/Firebase/config";
import { generateCode } from '../utils/codeGenerator';
import { FirebaseAuthContext } from '../components/Firebase/FirebaseAuthContext';
import MenuIcon from '../assets/images/menu-icon.svg';

const UserLinkScreen = ({ navigation }) => {

    const [userCode, setUserCode] = useState("");
    const [currentUserData, setUserData] = useState(null); 
    const { currentUser } = useContext(FirebaseAuthContext);
    const [connectType, setConnectType] = useState("");
    const [modalContentRender, setContentRender] = useState({
        showMethodsModal: true,
        showProvideCode: false,
        showInputCode: false
    });
    const [submitDisable, setSubmitDisable] = useState(false);

    useEffect(() => {

        if (currentUser) {

            firebase.firestore().collection("users").where("id", "==", currentUser.uid).get().then((querySnapshot) => {

                querySnapshot.forEach(e => {
                    // console.log(e.data());
                    setUserCode(e.data().connectCode);
                });

                // console.log(querySnapshot[0].data());
            });
        }

    }, []);


    const unselectedHP = {
        HPIcon: <HP_Btn style={styles.HPBtn} />,
        button: styles.connectButton,
        text: styles.connectButtonText,
    }

    const selectedHP = {
        HPIcon: <HP_BtnSelected style={styles.HPBtn} />,
        button: styles.selectedConnectButton,
        text: styles.selectedConnectButtonText,
    }

    const unselectedFF = {
        FamilyFriendIcon: <FamilyFriendBtn style={styles.FamilyFriendBtn} />,
        button: styles.connectButton,
        text: styles.connectButtonText,
    }

    const selectedFF = {
        FamilyFriendIcon: <FamilyFriendBtnSelected style={styles.FamilyFriendBtn} />,
        button: styles.selectedConnectButton,
        text: styles.selectedConnectButtonText,
    }

    const [connectButtonHP, setConnectButtonHP] = useState(unselectedHP);

    const [connectButtonFamilyFriend, setConnectButtonFamilyFriend] = useState(unselectedFF);

    const buttonSelect = (type) => {

        if (type === 'HEALTH_PRO') {
            setConnectButtonHP((state) => (selectedHP));
        } else if (type === 'FRIEND_FAMILY') {
            setConnectButtonFamilyFriend((state) => (selectedFF));
        }

        setConnectType(type);
        openModal();
    };

    const [isModalVisible, setModalVisible] = useState(false);

    // const [methodsShown, setMethodsShown] = useState(false);

    const buttonDeselect = () => {
        setConnectButtonHP(unselectedHP);
        setConnectButtonFamilyFriend(unselectedFF);
    };

    const closeModal = () => {
        buttonDeselect();
        setModalVisible(false);
        // setMethodsShown(false);
    };

    const openModal = () => {
        handleModalContent("METHODS");
        setModalVisible(true);
        // setMethodsShown(true);
    };

    const handleModalContent = (type) => {
        console.log("handled")
        if (type === "METHODS") {
            // modalContent = methodsModal;
            setContentRender(state => ({
                ...state,
                showMethodsModal: true
            }))
            setModalContent(methodsModal)
        } else if (type === "PROVIDE") {
            setModalContent(provideCodeModal);
            setContentRender(state => ({
                ...state,
                showMethodsModal: false,
                showProvideCode: true
            }))
        } else if (type === "INPUT") {
            setContentRender(state => ({
                ...state,
                showMethodsModal: false,  
                showProvideCode: false
            }))
            setModalContent(inputCodeModal);
        }
    };

    const useQR = () => {
        console.log("go to qr screen")
        closeModal();
        setTimeout(t => {
            navigation.navigate('QRScreen', { connecting: {
                connectUser: connectUser,
                connectMethod: "QR", 
                connectType: connectType
            }});
        });
    };

    const promptRefreshCode = () => {

        // console.log(NurseEllieLogo);
        Alert.alert("Refresh Code", "Are you sure?", [
            {
                text: 'Cancel',
            },
            {
                text: 'OK',
                onPress: () => refreshCode()
            },
        ], { cancelable: true });
        // alert("Are you sure you?");

    }

    const refreshCode = () => {
        let code = generateCode();
        firebase.firestore().collection("users").doc(currentUser.uid).update({
            connectCode: code
        }).then(() => {
            setUserCode(code);
        });

    };

    const modalGoBack = () => {
        console.log("modalGoBack()")

        if (methodsPressed) {
            handleModalContent("METHODS");
            setMethodsPressed(false);
            // setMethodsShown(false);  
        } else {
            closeModal();
            console.log(methodsPressed);
        }
    };

    const connectUser = async (connectMethod, connectType, data = null) => {
  
        let isFF = false;

        if (connectType == "FRIEND_FAMILY") {
            isFF = true;
        }

        await firebase.firestore().collection("users").where("connectCode", "==", (data) ? data : inputCode).get().then(async querySnapshot => {
            // results for user to be connected to
            let foundUser = false;
            let userID = null;  
            let user = null;
            
            querySnapshot.forEach(i => {
                foundUser = true;
                user = i.data();
                userID = i.id;   
            });

            if (foundUser === true) {

                let matches = [];

                await firebase.firestore().collection("users").doc(currentUser.uid).get().then(async doc => {
                   
                    if (doc.data()) {  
                        setUserData(doc.data());  
                        matches = doc.data().userLinks.filter(ref => user.userLinks.includes(ref));
                        // console.log(doc.data().userLinks, `user1`)
                        // console.log(user.userLinks, "user2")
                        // console.log(matches, `matches`);
                    }

                    if (matches.length > 0) {
                        alert("Already connected!");   
                    } else {
                        await firebase.firestore().collection("userlinks").add({
                            connectionType: connectType,
                            connectionMethod: connectMethod,
                            isFriendFamily: isFF,
                            user1: currentUser.uid,
                            user2: userID
                        }).then(async docRef => {   
                            let currentUserRefSet = false;
                            let secondUserRefSet = false;

                            // Setting current user ref
                            await firebase.firestore().collection("users").doc(currentUser.uid).update({
                                userLinks: firebase.firestore.FieldValue.arrayUnion(docRef.id)
                            }).then(() => {
                                currentUserRefSet = true;
                                console.log("Setted current user, userlinks array")
                            }).catch(error => {
                                console.log(error)
                            });

                            // Setting 2nd user ref
                            await firebase.firestore().collection("users").doc(userID).update({
                                userLinks: firebase.firestore.FieldValue.arrayUnion(docRef.id)
                            }).then(() => {
                                secondUserRefSet = true;
                                console.log("Setted second user, userlinks array")
                            }).catch(err => {
                                console.log(error)
                            });
                            
                            if (currentUserRefSet && secondUserRefSet) {
                                alert("Connected successfully!");
                            } else {
                                alert("Couldn't set references for either user!")
                            }

                        }).catch(error => {
                            alert("Failed to connect with user!");
                        });
                    }
                    // enables submit button
                    setSubmitDisable(false);
                });

            } else {
                alert("User not found!")
                // enables submit button
                setSubmitDisable(false);
            }

        }).catch((error) => {
            console.log(error);
        })

        // return "done";
    };
    const [methodsPressed, setMethodsPressed] = useState(false);

    const [inputCode, setInputCode] = useState("");

    const methodsModal = (
        <View>
            <View style={{ marginBottom: 15 }}>
                {/* select meth txxt */}
                <Text style={{ fontSize: 25, fontFamily: 'roboto-regular' }}>Select Method:</Text>
            </View>
            <View>
                {/* buttons */}
                <TouchableOpacity onPress={() => {
                    setMethodsPressed(true);
                    handleModalContent("PROVIDE");
                }
                } style={styles.methodBtn}>
                    <View style={{}}>
                        <Text style={styles.methodBtnText}>
                            Provide Code
                </Text>
                    </View>
                </TouchableOpacity>
                <TouchableOpacity onPress={useQR} style={styles.methodBtn}>
                    <View style={{}}>
                        <Text style={styles.methodBtnText}>
                            Scan QR Code
                </Text>
                    </View>
                </TouchableOpacity>
                <TouchableOpacity onPress={() => {
                    setMethodsPressed(true);
                    handleModalContent("INPUT")
                }} style={styles.methodBtn}>
                    <View style={{}}>
                        <Text style={styles.methodBtnText}>
                            Input Code
                </Text>
                    </View>
                </TouchableOpacity>
            </View>
        </View>
    );

    const provideCodeModal = (
        <View>
            <View style={{ alignItems: 'center' }}>
                <Text style={{ fontSize: 25, fontFamily: 'roboto-regular', marginBottom: "5%" }}>
                    Connect Code
                </Text>
                <QRCode
                    logo={NurseEllieLogo}
                    logoSize={65}
                    logoBackgroundColor='transparent'
                    size={250}
                    value={userCode}
                />
                <Text style={{ fontSize: 25, fontFamily: 'roboto-regular', marginTop: "5%" }}>
                    {userCode}
                </Text>
                <TouchableOpacity onPress={promptRefreshCode} style={styles.refreshCodeBtn}>
                    <View style={{}}>
                        <Text style={styles.methodBtnText}>
                            Refresh
                        </Text>
                    </View>
                </TouchableOpacity>
            </View>
        </View>
    );

    const inputCodeModal = (
        <View>
            <View style={{ alignItems: 'center', marginBottom: '10%' }}>
                <Text style={{ fontSize: 25, fontFamily: 'roboto-regular' }}>
                    Enter Connect Code:
                </Text>
            </View>
            <View>
                <TextInput
                    textAlign="center"
                    autoFocus={true}
                    autoCapitalize="characters"
                    style={[PatientStyles.textInput, {paddingTop: 8}]}
                    returnKeyType="done"
                    onSubmitEditing={Keyboard.dismiss} 
                    onChange={(event) => {
                        setInputCode(event.nativeEvent.text.trim());
                        // console.log(event.nativeEvent.text)
                    }}
                // value={inputCode}          
                >
                </TextInput>
                <TouchableOpacity disabled={submitDisable} onPress={() => {
                    // Disables button to prevent POSTing data more than once.
                    setSubmitDisable(true);
                    setTimeout(t=> {
                        connectUser("INPUT", connectType);
                    }, 250);
                    Keyboard.dismiss();
                }} style={[styles.methodBtn, { marginTop: "10%" }]}>
                    <View>
                        <Text style={styles.methodBtnText}>
                            Submit
                        </Text>
                    </View>
                </TouchableOpacity>
            </View>
        </View>
    );

    const [modalContent, setModalContent] = useState(methodsModal);

    return (
        <>
            <KeyboardAvoidingView style={{ flex: 1, }} behavior="padding" enabled>
                <Background />
                <TouchableOpacity style={PatientStyles.menuButton} onPress={() => navigation.openDrawer()}>
                    <MenuIcon />
                </TouchableOpacity>
                <Animatable.View style={[PatientStyles.drawer, {justifyContent: 'space-between'}]} animation="fadeInUpBig">
                    <View style={styles.screenHeader}>
                        <Text style={PatientStyles.headerFont}>
                            {`Stay\nConnected`}
                        </Text>
                        <NurseEllieConnectLogo height={75} style={styles.headerImage} />
                        {/* <Image style={styles.headerImage} source={NurseEllieConnectLogo} /> */}
                    </View>
                    <View style={styles.UserLinkScreenDescription}>
                        <Text style={styles.descriptionText}>
                            {`Connecting to a `}
                            <Text style={styles.underline}>
                                {`Health Professional`}
                            </Text>
                            {` will allow them to view your medication logs. And, will allow you to book appointments, report any symptoms, and connect with them.`}
                        </Text>
                    </View>
                    <View>
                        <Text style={styles.connectText}>
                            Connect to:
                    </Text>
                        <TouchableOpacity onPressIn={() => buttonSelect('HEALTH_PRO')} style={connectButtonHP.button}>
                            <View style={styles.buttonFormat}>
                                {connectButtonHP.HPIcon}
                                <Text style={connectButtonHP.text}>
                                    Health Professional
                            </Text>
                            </View>
                        </TouchableOpacity>
                        <TouchableOpacity onPressIn={() => buttonSelect('FRIEND_FAMILY')} style={connectButtonFamilyFriend.button}>
                            <View style={styles.buttonFormat}>
                                {connectButtonFamilyFriend.FamilyFriendIcon}
                                <Text style={connectButtonFamilyFriend.text}>
                                    Family Member / Friend
                            </Text>
                            </View>
                        </TouchableOpacity>
                    </View>
                    <Modal
                        // needs to be set to 0, otherwise it flickers when modal is closing.
                        backdropTransitionOutTiming={0}
                        isVisible={isModalVisible}
                        // onBackButtonPress={closeModal} 
                        onBackdropPress={closeModal}
                        onBackButtonPress={modalGoBack}
                    >
                        <View style={{ backgroundColor: 'white', padding: 25, paddingBottom: 55, borderRadius: 25 }}>
                            <View style={{ alignItems: 'flex-end' }}>
                                {/* exit btn */}
                                <TouchableOpacity onPressIn={() => { closeModal() }} pressRetentionOffset={{ bottom: 70, right: 70 }}>
                                    <CloseBtn />
                                </TouchableOpacity>
                            </View>
                            {/* Modal content */}
                            {/* Can't render using states for each JSX element, need to do this way to perserve state updates. */}
                            {modalContentRender.showMethodsModal ? methodsModal : 
                                !modalContentRender.showProvideCode ? inputCodeModal : provideCodeModal}
                        </View>
                    </Modal>
                </Animatable.View>
            </KeyboardAvoidingView>
        </>
    );
}

const styles = StyleSheet.create({
    refreshCodeBtn: {
        backgroundColor: '#42C86A',
        elevation: 3,
        borderRadius: 5,
        marginTop: 15,
        width: "90%"
    },
    methodBtn: {
        backgroundColor: '#42C86A',
        elevation: 3,
        borderRadius: 5,
        marginTop: 15,
    },
    methodBtnText: {
        fontFamily: 'roboto-medium',
        fontWeight: '900',
        fontSize: 17,
        alignSelf: 'center',
        color: 'white',
        textTransform: 'uppercase',
        margin: 10,
        letterSpacing: 2
    },
    buttonFormat: {
        flexDirection: 'row',
        marginLeft: 35
    },
    connectText: {
        fontSize: 20,
        color: 'grey'
    },
    HPBtn: {
        height: 35,
        resizeMode: 'contain',
        marginRight: 30
    },
    FamilyFriendBtn: {
        height: 24,
        resizeMode: 'contain',
        marginRight: 30
    },
    selectedConnectButton: {
        backgroundColor: '#42C86A',
        borderRadius: 20,
        elevation: 6,
        marginTop: 15,
        height: 70,
        justifyContent: 'center'
    },
    connectButton: {
        backgroundColor: '#fff',
        borderRadius: 20,
        elevation: 6,
        marginTop: 15,
        height: 70,
        justifyContent: 'center'
    },
    selectedConnectButtonText: {
        color: 'white',
        fontFamily: 'roboto-medium',
        fontSize: 17,
        alignSelf: 'center',
        marginLeft: 0
    },
    connectButtonText: {
        color: 'grey',
        fontFamily: 'roboto-medium',
        fontSize: 17,
        alignSelf: 'center',
        marginLeft: 0
    },
    UserLinkScreenDescription: {
        marginBottom: 85,
    },
    descriptionText: {
        fontFamily: 'roboto-medium',
        fontSize: 17,
        lineHeight: 23,
        color: 'grey'
    },
    underline: {
        textDecorationLine: 'underline'
    },
    screenHeader: {
        flexDirection: 'row',
    },
    headerImage: {
        flex: 1,
        right: 10,
    },
    UserLinkScreenDescriptionFont: {
        fontFamily: 'roboto-regular',
        fontSize: 12
    },
});

export default UserLinkScreen;