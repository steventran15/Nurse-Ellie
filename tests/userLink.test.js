import React from 'react';
import Enzyme, { shallow, mount } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import renderer from 'react-test-renderer';

Enzyme.configure({ adapter: new Adapter() });

import { firebase } from '../components/Firebase/config';
import { FirebaseAuthContext } from '../components/Firebase/FirebaseAuthContext';

import UserLinkScreen from '../screens/UserLinkScreen';
import QRScanner from '../components/QRScanner/qr-scanner';
import { Alert } from 'react-native';

// mandatory for esModule and default to be returned otherwise the QRCODE npm module won't be mocked properly.
// it will cause the test to fail. 
// NOTE: "When importing a default export, it's an instruction to import the property named default from the export object"
// Doc reference: https://jestjs.io/docs/en/jest-object#jestmockmodulename-factory-options
jest.mock('react-native-qrcode-svg', () => {
    return {
        __esModule: true,
        default: () => {
            return <div></div>;
        },
    };
});

// jest.mock('Alert', () => { 
//     return {
//         __esModule: true,
//         alert: jest.fn()
//     }
// });

const firestoreMock = {
    collection: jest.fn().mockReturnThis(),
    doc: jest.fn().mockReturnThis(),
    onSnapshot: jest.fn().mockResolvedValueOnce(),
    where: jest.fn().mockReturnThis(),
    get: jest.fn().mockResolvedValueOnce(),
    update: jest.fn().mockResolvedValueOnce() 
  };

export const FirebaseAuthProvider = ({ children }) => (
    <FirebaseAuthContext.Provider value={{ currentUser: { uid: '0nPQbdHvJ1YepznX7NruQ19OpAk2' } }}>
        {children}
    </FirebaseAuthContext.Provider>
);

const navigationProp = {
    navigate: jest.fn(),
    openDrawer: jest.fn(),
    goBack: jest.fn()
};

const routeProp = {
    params: {
        connecting: {
            // connectUser: connectUser,
            connectUser: jest.fn().mockReturnThis(),
            connectMethod: "QR",
            // connectType: connectType
            connectType: 'HEALTH_PRO'
        },
    },
};

// Mocks the functions required to use firestore();
jest.spyOn(firebase, 'firestore').mockImplementationOnce(() => firestoreMock);

// const setConnectButtonHP = jest.fn();
// const setConnectType = jest.fn();
// const handleButtonSelect = jest.spyOn(React, "useState").mockImplementation(connectButtonHP => [connectButtonHP, setConnectButtonHP]);
// const handleSetConnectType = jest.spyOn(React, "useState").mockImplementation(connectType => [connectType, setConnectType]);

// Define wrapper as a global in this describe block, so it can be accessed in each it() test case.
const wrapper = mount(
    <FirebaseAuthProvider>
        <UserLinkScreen navigation={navigationProp} />
    </FirebaseAuthProvider>
);



describe('USER LINK SCREEN - PROVIDE CODE', () => {
    
     it('USER LINK SCREENS RENDERS CORRECTLY', () => {
        
// const userLinkWrapper = renderer.create(
//     <FirebaseAuthProvider>
//         <UserLinkScreen navigation={navigationProp} />
//     </FirebaseAuthProvider>
// );
// expect(userLinkWrapper.toJSON()).toMatchSnapshot();

        expect(wrapper.html()).toMatchSnapshot();
    })


    it('SETS USER CODE IN INITIAL RENDER', ()=> {
        const setUserCode = jest.fn(); 
        const handleSetUserCode = jest.spyOn(React, "useState").mockImplementation(userCode => [userCode, setUserCode]);

        expect(setUserCode).toBeTruthy();
    });

    // Tests if the button is clicked.
    it('SELECT HEALTH PROFESSIONAL BUTTON', () => {

        // Looks for all TouchableOpacity button nodes, finds the one with the prop testID then clicks.
        wrapper.find('TouchableOpacity').forEach(node => {
            if (node.prop("testID") === "hp_button") {

                // calls function but doesnt actually change the button color.
                // node.props().onPress();
                node.simulate('click');
            }
        });
        
        // Once clicked, if button style is changed then test is PASSED.
        wrapper.find('TouchableOpacity').forEach(node => {
            if (node.prop("testID") === "hp_button") {
                expect(node.prop("style").backgroundColor).toEqual('#42C86A');
            }
        });

    });

    it('METHODS MODAL CONTENT LOADED', () => {
        // If length of nodes for methodsModal == 1 that mean it is rendered after the button is clicked.
        // Otherwise the length would be 0 if the button isn't clicked.
        expect(wrapper.findWhere(node => node.prop("testID") === "methodsModal").hostNodes()).toHaveLength(1);
    });
    
    it('METHODS MODAL OPENS', () => {
        // expect(wrapper.findWhere(node=> node.prop('testID') === "modal").hostNodes()).toHaveLength(1);
        // Looks for the modal node, then grabs the first node and checks the visisble prop to see if it is true. 
        // If true, that means modal is opened.
        // NOTE: there are 2 nodes the first node is the proper node and the 2nd one does not have all the data/props for the modal node. 
        expect(wrapper.findWhere(node => node.prop('testID') === "modal").hostNodes().first().props().visible).toEqual(true);
    });

    it('SELECT PROVIDE CODE BUTTON', () => {
        // If length of nodes for methodsModal == 1 that mean it is rendered after the button is clicked.
        // Otherwise the length would be 0 if the button isn't clicked.
        // expect(wrapper.findWhere(node => node.prop("testID") === "provideCodeButton").hostNodes()).toHaveLength(1);
        wrapper.findWhere(node => node.prop("testID") === "provideCodeButton").hostNodes().simulate('click');
    });

    it('PROVIDE CODE MODAL CONTENT LOADED', () => {
        // If length of nodes for methodsModal == 1 that mean it is rendered after the button is clicked.
        // Otherwise the length would be 0 if the button isn't clicked.
        expect(wrapper.findWhere(node => node.prop("testID") === "provideCodeModal").hostNodes()).toHaveLength(1);
    });

    it('QR HAS DATA AND IS SHOWN', () => {
        
        // (wrapper.findWhere(node => node.prop('testID') === "qrcode")).forEach(node=> {
        //     console.log(node.props())
        // }); 
        // the value prop passed to QRCode component should be the user code. If the data was fetched in the initial render of the screen.
        expect(wrapper.findWhere(node => node.prop('testID') === "qrcode").props().value).toEqual("");
           
    });
     
    it('REFRESH CODE', () => {
        // Called firebase spyOn again for this refresh code test because it doesn't recongize that update() has been mocked in the beginning.
        jest.spyOn(firebase, 'firestore').mockImplementationOnce(() => firestoreMock);
        jest.spyOn(Alert, 'alert')
        
        // (wrapper.findWhere(node => node.prop('testID') === "userCode").hostNodes()).forEach(node => {
        //     console.log(node.props())
        // })  
        wrapper.findWhere(node=> node.prop('testID') === "refreshButton").hostNodes().simulate('click');
        expect(Alert.alert).toHaveBeenCalled();
        (Alert.alert.mock.calls[0][2][1].onPress());
        expect(wrapper.findWhere(node => node.prop('testID') === "userCode").hostNodes().props().children).toEqual("");
    });

    it('CLOSE MODAL', () => {
        // expect(wrapper.findWhere(node=> node.prop('testID') === "closeButton").hostNodes().forEach(node=> {
        //     console.log(node.props())
        // })).toHaveLength(1)
        // const setModalVisible = jest.fn();
        // const handleSetModal = jest.spyOn(React, 'useState').mockImplementation((isModalVisible) => [isModalVisible, setModalVisible]);
        // wrapper.findWhere(node => node.prop('testID') === "closeButton").hostNodes().simulate('click');
        // expect(setModalVisible).toBeTruthy();
        // expect(wrapper.findWhere(node => node.prop('testID') === "modal").hostNodes().first().props().visible).toEqual(false);
    });

});

describe('USER LINK SCREEN - INPUT CODE', () => {
    it('SELECT INPUT CODE BUTTON', () => {
        wrapper.find('TouchableOpacity').forEach(node => {
            if (node.prop("testID") === "hp_button") {
                node.simulate('click');
            }
        });

        wrapper.findWhere(node => node.prop("testID") === "inputCodeButton").hostNodes().simulate('click');
    });

    it('INPUT CODE MODAL CONTENT LOADED', () => {
        expect(wrapper.findWhere(node => node.prop("testID") === "inputCodeModal").hostNodes()).toHaveLength(1);
    });

});

describe('QR SCANNER', () => {

    it('QRSCanner renders correctly', () => {
        // Needed to mock setTimeout();
        jest.useFakeTimers();
        
        const qrWrapper = renderer.create(
            <QRScanner {...navigationProp} {...routeProp}/>
        );
        expect(qrWrapper.toJSON()).toMatchSnapshot();

        // let qrWrapper = mount(<QRScanner {...navigationProp} {...routeProp}/>);
    });

    it('CONNECT USER', () => { });

});