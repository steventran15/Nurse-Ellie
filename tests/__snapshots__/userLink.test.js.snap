// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`QR SCANNER QRSCanner renders correctly 1`] = `
<View
  style={
    Object {
      "backgroundColor": "black",
      "flex": 1,
    }
  }
>
  <View
    style={
      Object {
        "alignItems": "flex-end",
        "zIndex": 99,
      }
    }
  >
    <svg
      data-file-name="SvgFlashOff"
      onPress={[Function]}
      style={
        Object {
          "marginRight": 10,
          "marginTop": "2%",
        }
      }
      width={56.25}
    />
  </View>
  <View
    style={
      Object {
        "alignItems": "center",
        "zIndex": 99,
      }
    }
  >
    <Text
      style={
        Object {
          "color": "white",
          "fontFamily": "roboto-regular",
          "fontSize": 67.5,
          "marginTop": "5%",
          "textAlign": "center",
          "width": "100%",
        }
      }
    >
      Scan Code
    </Text>
    <svg
      data-file-name="SvgQrSquare"
      style={
        Object {
          "height": 525,
          "marginBottom": "20%",
          "marginTop": "20%",
          "width": 525,
        }
      }
    />
    <Text
      onPress={[Function]}
      style={
        Object {
          "color": "white",
          "fontFamily": "roboto-regular",
          "fontSize": 67.5,
          "marginTop": "10%",
          "textAlign": "center",
          "width": "100%",
        }
      }
    >
      Cancel
    </Text>
  </View>
</View>
`;

exports[`USER LINK SCREEN - PROVIDE CODE USER LINK SCREENS RENDERS CORRECTLY 1`] = `
"<view><view style=\\"flex: 1; justify-content: center; background-color: rgb(66, 200, 106);\\"><view style=\\"height: 500px; width: 500px; border-radius: 250px; background-color: rgba(62, 185, 99, 0.8); position: absolute; right: -375px; top: 667px;\\"></view><view style=\\"height: 60px; width: 60px; border-radius: 30px; background-color: rgba(255, 255, 255, 0.3); position: absolute; left: 10px; top: 30px;\\"></view><view style=\\"height: 250px; width: 250px; border-radius: 125px; background-color: rgba(255, 255, 255, 0.3); position: absolute; left: 150px; top: 222.33333333333334px;\\"></view><view style=\\"height: 600px; width: 600px; border-radius: 300px; background-color: rgba(255, 255, 255, 0.3); position: absolute; left: 300px; top: -333.5px;\\"></view><view style=\\"height: 350px; width: 350px; border-radius: 175px; background-color: rgba(255, 255, 255, 0.3); position: absolute; left: -250px; top: 667px;\\"></view></view><view style=\\"position: absolute; right: 30px; top: 40px; opacity: 1;\\" focusable=\\"true\\"><svg data-file-name=\\"SvgMenuIcon\\"></svg></view><view style=\\"flex: 4; background-color: rgb(255, 255, 255); border-top-left-radius: 30px; border-top-right-radius: 30px; position: absolute; width: 750px; height: 1133.8999999999999px; top: 200.1px; justify-content: space-between; opacity: 0; transform: [object Object];\\"><view style=\\"flex-direction: row;\\"><text style=\\"font-family: roboto-regular; font-size: 32px; font-weight: 100;\\">Stay
Connected</text><svg height=\\"75\\" style=\\"flex: 1; right: 10px;\\" data-file-name=\\"SvgEllieConnectLogo\\"></svg></view><view style=\\"margin-bottom: 85px;\\"><text style=\\"font-family: roboto-medium; font-size: 17px; line-height: 23; color: grey;\\">Connecting to a <text style=\\"text-decoration-line: underline;\\">Health Professional</text> will allow them to view your medication logs. And, will allow you to book appointments, report any symptoms, and connect with them.</text></view><view><text style=\\"font-size: 20px; color: grey;\\">Connect to:</text><view style=\\"background-color: rgb(255, 255, 255); border-radius: 20px; elevation: 6px; margin-top: 15px; height: 70px; justify-content: center; opacity: 1;\\" testid=\\"hp_button\\" focusable=\\"true\\"><view style=\\"flex-direction: row; margin-left: 35px;\\"><svg style=\\"height: 35px; margin-right: 30px;\\" data-file-name=\\"SvgNurseUnselectedIcon\\"></svg><text style=\\"color: grey; font-family: roboto-medium; font-size: 17px; align-self: center; margin-left: 0px;\\">Health Professional</text></view></view><view style=\\"background-color: rgb(255, 255, 255); border-radius: 20px; elevation: 6px; margin-top: 15px; height: 70px; justify-content: center; opacity: 1;\\" focusable=\\"true\\"><view style=\\"flex-direction: row; margin-left: 35px;\\"><svg style=\\"height: 24px; margin-right: 30px;\\" data-file-name=\\"SvgFamilyfriendUnselectedIcon\\"></svg><text style=\\"color: grey; font-family: roboto-medium; font-size: 17px; align-self: center; margin-left: 0px;\\">Family Member / Friend</text></view></view></view><modal animationtype=\\"none\\" testid=\\"modal\\" swipethreshold=\\"100\\" scrolloffset=\\"0\\" scrolloffsetmax=\\"0\\" supportedorientations=\\"portrait,landscape\\"><view style=\\"position: absolute; top: 0px; bottom: 0px; left: 0px; right: 0px; opacity: 0; background-color: black; width: 750px; height: 1334px;\\" focusable=\\"true\\"></view><view style=\\"margin: 37.5px; transform: [object Object]; flex: 1; justify-content: center;\\" pointer-events=\\"box-none\\" testid=\\"modal\\" swipethreshold=\\"100\\" scrolloffset=\\"0\\" scrolloffsetmax=\\"0\\" supportedorientations=\\"portrait,landscape\\"><view testid=\\"modalContent\\" style=\\"background-color: white; padding: 25px 25px 55px 25px; border-radius: 25px;\\"><view style=\\"align-items: flex-end;\\"><view style=\\"opacity: 1;\\" testid=\\"closeButton\\" focusable=\\"true\\"><svg data-file-name=\\"SvgCloseButton\\"></svg></view></view><view testid=\\"methodsModal\\"><view style=\\"margin-bottom: 15px;\\"><text style=\\"font-size: 25px; font-family: roboto-regular;\\">Select Method:</text></view><view><view style=\\"background-color: rgb(66, 200, 106); elevation: 3px; border-radius: 5px; margin-top: 15px; opacity: 1;\\" testid=\\"provideCodeButton\\" focusable=\\"true\\"><view><text style=\\"font-family: roboto-medium; font-weight: 900; font-size: 17px; align-self: center; color: white; text-transform: uppercase; margin: 10px; letter-spacing: 2px;\\">Provide Code</text></view></view><view style=\\"background-color: rgb(66, 200, 106); elevation: 3px; border-radius: 5px; margin-top: 15px; opacity: 1;\\" focusable=\\"true\\"><view><text style=\\"font-family: roboto-medium; font-weight: 900; font-size: 17px; align-self: center; color: white; text-transform: uppercase; margin: 10px; letter-spacing: 2px;\\">Scan QR Code</text></view></view><view style=\\"background-color: rgb(66, 200, 106); elevation: 3px; border-radius: 5px; margin-top: 15px; opacity: 1;\\" testid=\\"inputCodeButton\\" focusable=\\"true\\"><view><text style=\\"font-family: roboto-medium; font-weight: 900; font-size: 17px; align-self: center; color: white; text-transform: uppercase; margin: 10px; letter-spacing: 2px;\\">Input Code</text></view></view></view></view></view></view></modal></view></view>"
`;
